### DISCUSSION TITLE
Does uWebSockets.js support http.request, or a node-style fetch?

### DISCUSSION THREAD

**Discussion by boulderwebdev** [uNetworking/uWebSockets.js#390] [Q&A] [answered]
I've tried making an http request to a private http authentication server during the `onUpgrade` call for a websocket server using various methods but keep getting an `Aborted (core dumped)` error in the command line. It seems like the uWebSockets.js module strips node from making it's own sockets, or node is unable to deal with the uSockets. Is this correct? If so, then this issue is a feature request for getting http support with uWebSockets.js. Otherwise, does anyone have some example code which allows me to make http requests? I have

```
node v10.16.0
github:uNetworking/uWebSockets.js#v18.8.0
node-fetch 2.6.1
```

installed and have tried making requests using both `node-fetch` and `http.request` (although I'm guessing the former depends on the latter).

**Comment by hst-m** [comment]
I just tried node http and it works fine, you are probably doing something wrong, post your code here. I use this [bent](https://github.com/mikeal/bent) library for all my requests, it is newish based on promises. Did you try newer Node version, v10 is old they at 15 now
```javascript
const getJson=require('bent')('json')
const obj=await getJson(url)
```

**Comment by ghost** [comment]
> It seems like the uWebSockets.js module strips node from making it's own sockets, or node is unable to deal with the uSockets. Is this correct?

No this is not correct. You are doing something else - most probably passing wrong types or wrong number of arguments to functions with relaxed type checking.

How does your code look like?

**Comment by boulderwebdev** [comment]
@alexhultman Here's the short snippet:

```js
import fetch from "node-fetch";
import * as http from "http";

export class UWebSocketsTracker {
  private async isAuthenticated(request: HttpRequest): Promise<boolean> {
    const headers = {
      "X-Auth": "some-val"
    };
    const url = "http://localhost:9000";
    // request code
  }
}
```

where the `code` section is one of

```js
try {
  const resp = await fetch(url, headers);
  console.log(resp);
  return resp.ok;
} catch (e) {
  console.log("exception", e);
  return false;
}
```

or

```js
return fetch(url, headers).then(resp => {
  return resp.ok;
}).catch(e => {
  console.log(e);
  return false;
})
```

----

```js
private readonly onUpgrade = (
  res: HttpResponse, req: HttpRequest, context: us_socket_context_t
): void => {
  const url = req.getUrl();
  const secWebSocketKey = req.getHeader('sec-websocket-key');
  const secWebSocketProtocol = req.getHeader('sec-websocket-protocol');
  const secWebSocketExtensions = req.getHeader('sec-websocket-extensions');

  this.isAuthenticated(req).then(isAuthenticated => {
    if (isAuthenticated) {
      res.upgrade(
        { url: url },
        secWebSocketKey,
        secWebSocketProtocol,
        secWebSocketExtensions,
        context
      );
    } else {
      throw new Error("not authenticate");
    }
  }).catch(e => {
    res.writeStatus("400 Not Authorized").end("not authenticated");
  })
}
```

**Comment by ghost** [comment]
Well I want to see your use of uWebSockets.js, not some entirely irrelevant 😉

**Comment by hst-m** [comment]
@boulderwebdev simplify to post full file of an example that throws error, so can see what you are doing wrong. Also try updating to uWS.js v18.10 which includes some type checking, it would print out error for you

@alexhultman you need to update Readme for new versions, it still says latest install is v18.8 which is what he using

**Comment by boulderwebdev** [comment]
@hst-m Let me write up an example real quick

@alexhultman I updated the code above, but will create a test case so we can all see what's exactly going on.

Also, here's the full error message

```sh
Error: Returning from a request handler without responding or attaching an abort handler is forbidden!
terminate called without an active exception
Aborted (core dumped)
```

**Comment by ghost** [comment]
Okay so there you have it

**Comment by hst-m** [comment, marked as answer]
> Error: Returning from a request handler without responding or attaching an abort handler is forbidden!
terminate called without an active exception

@boulderwebdev Why did you not include that in issue? that is your problem, you need to respond to the request or attach onAborted handler before doing any async, see [example](https://github.com/uNetworking/uWebSockets.js/blob/master/examples/AsyncFunction.js)

**Comment by boulderwebdev** [comment]
Ah, thanks for the pointer @hst-m

**Comment by hst-m** [comment]
It is helpful to post the full error message if you are going to create an issue for this error :P
